### PR TITLE
Ensure the async driver applies connection string settings

### PR DESCRIPTION
**Pull request by rozza**
JAVA-2270

### CODE REVIEW

**Comment by rozza** [comment]
Refs mongodb/mongo-java-driver#363

File: driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy [hunk 1]
```diff
@@ -118,4 +126,71 @@ class MongoClientsSpecification extends FunctionalSpecification {
         cleanup:
         client?.close()
     }
+
+    @Unroll
+    def 'should correct parse read preference for #readPreference'() {
```
Comment (jyemin) [inline driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy]: It's not really about parsing.   To be consistent with other test method names, how about:

    def 'should apply read preference from connection string to settings'

File: driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy [hunk 2]
```diff
@@ -118,4 +126,71 @@ class MongoClientsSpecification extends FunctionalSpecification {
...
+
+        then:
+        client.settings.getReadPreference() == readPreference
+
+        cleanup:
+        client?.close()
+
+        where:
+        uri                                                              | readPreference
+        'mongodb://localhost/?readPreference=secondaryPreferred'         | secondaryPreferred()
```
Comment (jyemin) [inline driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy]: All these different read preferences is more of a test of the ConnectionString class than of the MongoClients class.  It's sufficient to test just one.

File: driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy [hunk 3]
```diff
@@ -118,4 +126,71 @@ class MongoClientsSpecification extends FunctionalSpecification {
...
+                '&readPreferenceTags=dc:ny,rack:1' +
+                '&readPreferenceTags=dc:ny' +
+                '&readPreferenceTags='                                   | secondaryPreferred([
+                                                                            new TagSet(asList(new Tag('dc', 'ny'), new Tag('rack', '1'))),
+                                                                            new TagSet(asList(new Tag('dc', 'ny'))),
+                                                                            new TagSet()])
+    }
+
+    @Unroll
+    def 'should correct parse read concern for #readConcern'() {
```
Comment (jyemin) [inline driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy]: As above

File: driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy [hunk 4]
```diff
@@ -118,4 +126,71 @@ class MongoClientsSpecification extends FunctionalSpecification {
...
+
+        then:
+        client.settings.getReadConcern() == readConcern
+
+        cleanup:
+        client?.close()
+
+        where:
+        uri                                               | readConcern
+        'mongodb://localhost/'                            | ReadConcern.DEFAULT
```
Comment (jyemin) [inline driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy]: As above

File: driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy [hunk 5]
```diff
@@ -118,4 +126,71 @@ class MongoClientsSpecification extends FunctionalSpecification {
...
+        'mongodb://localhost/?readConcernLevel=local'     | ReadConcern.LOCAL
+        'mongodb://localhost/?readConcernLevel=majority'  | ReadConcern.MAJORITY
+    }
+
+    @Unroll
+    def 'should correctly parse different #writeConcern'() {
```
Comment (jyemin) [inline driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy]: As above

File: driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy [hunk 6]
```diff
@@ -118,4 +126,71 @@ class MongoClientsSpecification extends FunctionalSpecification {
...
+        when:
+        def client = MongoClients.create(uri)
+
+        then:
+        client.settings.getWriteConcern() == writeConcern
+
+        cleanup:
+        client?.close()
+
+        where:
```
Comment (jyemin) [inline driver-async/src/test/functional/com/mongodb/async/client/MongoClientsSpecification.groovy]: As above

**Comment by rozza** [comment]
Updated @jyemin

**Comment by jyemin** [comment]
LGTM

**Comment by rozza** [comment]
Merged to 3.3.x and master

**Pull request closed without merging**